base/launcher/src/web_server.cc

```diff
@@ -24,6 +25,108 @@ using namespace motis::module;
 
 namespace motis::launcher {
 
+std::string encode_msg(msg_ptr const& msg, bool const binary) {
+  std::string b;
+  if (binary) {
+    b = std::string{reinterpret_cast<char const*>(msg->data()), msg->size()};
+  } else {
+    b = msg->to_json();
+  }
+  return b;
+}
+
+msg_ptr build_reply(int const id, msg_ptr const& res,
+                    std::error_code const& ec) {
+  msg_ptr m = res;
+  if (ec) {
+    m = make_error_msg(ec);
+  } else if (!res) {
+    m = make_success_msg();
+  }
+  m->get()->mutate_id(id);
+  return m;
+}
+
+msg_ptr decode_msg(std::string const& req_buf, bool const binary) {
+  if (binary) {
+    return make_msg(req_buf.data(), req_buf.size());
+  } else {
+    return make_msg(req_buf, true);
+  }
+}
+
+struct ws_client : public client,
+                   public std::enable_shared_from_this<ws_client> {
+  ws_client(boost::asio::io_service& ios, net::ws_session_ptr session,
+            bool binary)
+      : ios_{ios}, session_{std::move(session)}, binary_{binary} {}
+
+  ws_client(ws_client const&) = delete;
+  ws_client(ws_client&&) = delete;
+  ws_client& operator=(ws_client const&) = delete;
+  ws_client& operator=(ws_client&&) = delete;
+
+  ~ws_client() override = default;
+
+  void set_on_msg_cb(std::function<void(msg_ptr const&)>&& cb) override {
+    if (auto const lock = session_.lock(); lock) {
+      lock->on_msg([this, cb = std::move(cb)](std::string const& req_buf,
+                                              net::ws_msg_type const type) {
+        msg_ptr err;
+        int req_id = 0;
+
+        try {
+          auto const req =
+              decode_msg(req_buf, type == net::ws_msg_type::BINARY);
+          req_id = req->get()->id();
+          return cb(req);
+        } catch (std::system_error const& e) {
+          err = build_reply(req_id, nullptr, e.code());
+        } catch (std::exception const& e) {
+          err = make_unknown_error_msg(e.what());
+        } catch (...) {
+          err = make_unknown_error_msg("unknown");
+        }
+        err->get()->mutate_id(req_id);
+
+        if (auto const lock = session_.lock(); lock) {
+          lock->send(err->to_json(), type,
```

Comment:
```suggestion
          lock->send(encode_msg(err, type == net::ws_msg_type::BINARY), type,
```

(Or using `binary_` like in `send` (but `type` should be consistent))